AddressMapper client

eduardonunesp opened the pull request.

enlight commented on src/address_mapper.ts:
```diff
@@ -49,19 +50,22 @@ export class AddressMapper {
+      [
+        { type: 'address', value: from.local.toString().slice(2) },
+        { type: 'address', value: to.local.toString().slice(2) }
+      ].join('')
```
This `join()` doesn't make sense.

eduardonunesp replied:
@enlight the Sha3 result should be the `from` and `to` concatenated isn't?

enlight commented on src/address_mapper.ts:
```diff
@@ -49,19 +50,22 @@ export class AddressMapper {
-  }
-
-  async addIdentityMapping(from: Address, to: Address, privateKey: Uint8Array) {
+  async addIdentityMapping(from: Address, to: Address, web3: Web3, ethAddress: string) {
```
A `Web3Signer` instance should be passed in as a parameter, you shouldn't create it in here.

enlight replied:
Also this function should be named `addIdentityMappingAsync`

enlight commented on src/address_mapper.ts:
```diff
@@ -0,0 +1,72 @@
+  private _privateKey: Uint8Array
+  private _addressMapperContract?: Contract
+
+  constructor(client: Client, privateKey: Uint8Array) {
```
Get rid of this private key too, you only need the caller address.

enlight commented on src/plasma-cash/solidity-helpers.ts:
```diff
@@ -31,10 +31,14 @@ export class Web3Signer {
+    const v = ethutil.toBuffer(
+      parseInt(sig.substring(128, 130), 16) + (isMetamask ? 0 : 27)
+    ) as Buffer
+    const mode = ethutil.toBuffer(isMetamask ? 3 : 1) as Buffer // mode = geth
```
`mode == 3` isn't going to work, it needs to be 1.

As for the `isMetamask` bit, let's just do this:
```js
let v = parseInt(sig.substring(128, 130), 16)
if (v === 0 || v ===1) {
  v += 27
}
```

eduardonunesp commented:
Removed the private key from constructor, passing the address, also added a factory static method to create the contract and return `AddressMapper`  ready to use
